components/resource_control/src/resource_limiter.rs

```diff
@@ -154,29 +157,33 @@ impl QuotaLimiter {
         )
     }
 
-    fn consume(&self, value: u64) -> Duration {
-        if value == 0 {
+    fn consume(&self, value: u64, wait: bool) -> Duration {
+        if value == 0 && self.limiter.speed_limit().is_infinite() {
             return Duration::ZERO;
         }
-        let dur = self.limiter.consume_duration(value as usize);
-        if dur != Duration::ZERO {
+        let mut dur = self.limiter.consume_duration(value as usize);
+        if !wait {
```

Comment:
Can `consume` still return the duration but we handle it as `debet` in the LimitFuture?  then no need wait flag.

Reply:
Then when to pay back the debt? When wait is false, it means the future is ready and after this call the future is returned/dropped, so the future has no chance to pay the debt.

Reply:
If I'm not misunderstanding, it will be paid back in the next future's first poll. this flag seems useless in the `consume` function.

Reply:
No. The next futures's first poll will call `consume(0, ture)` and it may return a non-zero wait duration. So if the last poll also returns a non-zero duration, that means the wait duration is recorded twice in the `TASKS_WAIT_DURATION` metric, this may lead to overestimation of the expected cpu time

Reply:
Maybe we can move the `TASKS_WAIT_DURATION` to the caller, in this way, the resource limiter won't be responsible to manage the total wait time and the `wait` flag won't be need.

Reply:
yes, I think it's okay to record in the caller, actually, the wait behavior do by the caller.

Reply:
It's hard to move the quota-limit wait duration(total_wait_dur_us) to the caller side(that is, in LimitedFuture). Currently, we have two different types of ResourceLimiter, background limiter and priority limiter. To maintain these different types of wait time is not easy. And further more, we have cpu and IO quota-limiter, as we only return a single waiting duration, caller can't determine the source of this duration, e.g., we handle the resource quota by cpu currently, if io is the bottleneck, it can overestimate the cpu requirement and vice verse.